### ISSUE TITLE
Provide Elliptic Integral of the Third Kind

### ISSUE THREAD

**Issue by sergiopasra** [scipy/scipy#4452]
The scipy.special module provides elliptic integral of the first and second kind, but the elliptical integral of the third kind is missing.
Sympy implements it: http://docs.sympy.org/latest/modules/functions/special.html#sympy.functions.special.elliptic_integrals.elliptic_pi

Further references: http://mathworld.wolfram.com/EllipticIntegraloftheThirdKind.html

**Comment by ewmoore** [comment]
Doesn't sympy just call mpmath to evaluate things numerically?

Boost has an implementation, but using it would require porting it away
from the boost infrastructure for us to use it.

**Comment by sergiopasra** [comment]
This is the code in sympy.

https://github.com/sympy/sympy/blob/master/sympy/functions/special/elliptic_integrals.py#L253

Boost is using the algorithms described here:

http://arxiv.org/abs/math/9409227

Or something very similar.

Scipy wraps cephes, but it doesn't have the third kind elliptical

**Comment by gmcastil** [comment]
Looks like a cool addition, I'll look into this.

**Comment by gmcastil** [comment]
I've been reading up on the algorithms described in [this](http://arxiv.org/abs/math/9409227) and on Carlson forms of elliptic integrals in general.  I was thinking to start with writing some tests for proper computation of R<sub>F</sub> and R<sub>J</sub> (the paper gives numerical checks) and then implementing something like `special.ellippi()` and `ellippiinc()` in terms of R<sub>F</sub> and R<sub>J</sub>.

I'm still working on understanding the algorithm - any other advice would be welcome.

**Comment by ewmoore** [comment]
Using R_F and R_J is the right approach.  There is an implementation of R_F for real inputs here: https://github.com/scipy/scipy/blob/master/scipy/special/cephes/ellik.c#L161. It's not standalone or totally complete, but it might get you started. 

Test data sets, from the boost project, are also already in scipy for R_F, R_J and ellippi. Look under scipy/special/tests/data/boost.

**Comment by sergiopasra** [comment]
I started to write some code for Rc, Rf and Rj here: https://github.com/sergiopasra/pylliptical 
I used the algorithms from http://arxiv.org/abs/math/9409227, it's cool that they work for complex numbers too.

**Comment by gmcastil** [comment]
I am going to generalize this a bit into a larger effort to simply implement the Carlson symmetric forms, and then implement elliptic integrals of the third kind in terms of those.  Sort of surprised that these have not been implemented directly yet.  I don't know how much use there is for these, but I should think these could be a welcome addition.

@sergiopasra I should have something up on Github by the end of the week - feel free to check it out.

**Comment by gmcastil** [comment]
@ewmoore I think I was able to figure out the Boost datasets (at least the ones for R_C), but they do not appear to have any complex inputs.  Any suggestion on where to find a more comprehensive dataset?

Additionally, I've been able to implement the R_C and R_J algorithms and they produce correct values for all the checks that are given in the previously linked paper, but I'd still like to fire them at larger sets of problems.

Thanks.

**Comment by gmcastil** [comment]
I've been working on this addition a fair bit the last few days.  For now it's just a separate repo here https://github.com/gmcastil/ellyptic.  I've also added a branch for developing tests for it as well - it uses the Boost library as well as some complex valued discrete examples from the referenced paper describing the algorithms.

So far, everything looks pretty good from my perspective - I should get R_J and R_D done tomorrow and then I'll look at implementing Legendre forms in terms of the Carlson forms.  Once everything is in order, I'll post on the scipy-devel mailing list about possible inclusion of these in addition to, or as replacements for the existing Legendre forms.

Any suggestions or additional advice is much appreciated; this is my first non-trivial contribution to the SciPy community, so I'm grateful for any guidance.  Thanks.

**Comment by gmcastil** [comment]
The remaining two Carlson forms are working at this point, but there is still a bug I'm trying to figure out, where some things are not working out like they should.  The `master` branch is currently just R_F and R_C (which is needed to approximate R_J) and it works and passes all the discrete tests as well as the Boost set.  The `remaining-funcs` branch is where I'm doing the development of the last two.  If anyone wants to take a look and offer suggestions, much appreciated.  Thanks.

**Comment by rgommers** [comment]
I'm not really a special functions guy, so no comments on the algorithms. I did have a look at the code and the paper though, looks relatively straightforward to match those for someone new to the code.

Would be good to add the proper citation for the paper in the public ``ellip*`` functions, which I think is

    B. C. Carlson, "Numerical computation of real or complex elliptic integrals",
    Numerical Algorithms vol. 10, pp. 13-26, 1995.

**Comment by ewmoore** [comment]
I think you should drop the `rtol` parameter and calculate to machine precision at all times.   There are several reasons for this.  Firstly, even if written in Python, these functions should act like ufuncs which do not support such an argument, and secondly, a default that isn't machine precision is a bug in waiting, since no one will change the default.  I know Carlson uses something like your `3e-4` in his papers but the algorithms are much more capable than that.   

`R_F(x, y, z)` is symmetric in all of `x`, `y`, and `z`.  So only checking `y == z` for the special case isn't really enough.  `x == y` or `x == z` are also equal to a call to `R_C`.  Checking for exact equality is fine here, but generally the code in R_F ought to be able to handle that case too, since  `R_F(x, x+eps, z)` will fail that check (as it should) but likely give a very similar value.

What happens for the special values such as `NaN` or `Inf`?  These should be tested.

What are the branch cuts for these?  http://fredrik-j.blogspot.com/2010/06/incomplete-elliptic-integrals-complete.html makes it look potentially tricky to get this right for everything.

Using these to add support for complex numbers to the existing elliptic integrals will require integrating them with the rest of the the special package.  This means that you'll need to port this C or cython (or whatever).

Looks like you've got a good start here.

**Comment by gmcastil** [comment]
Thanks for the advice @ewmoore.  I'll look into addressing the items you mentioned.  I had not done any type checking yet because I wanted to make sure that it was working first.

I had thought about the case where two arguments were nearly identical, but wasn't sure what the best method for doing something like that was.  Agree with you about testing NaN and +/- Inf.

For branch cuts, I will re-read some of the material to try to see what branch cuts had been implicitly selected for each function and maybe check out the `mpmath` documentation for some guidance on the subject.  I agree that they should be explicitly defined.

Ultimately, I would like to do more or less what was done for `mpmath` - implementing all three of the Legendre forms in terms of Carlson symmetric forms and making both sets available in the API.  Nominally thinking to use the Boost dataset that ships with SciPy currently, along with the special checks referenced by Carlson.  I've not done anything using Cython, so this is good opportunity for me to do so.

This will be the most sophisticated contribution I've made to the toolstack - I have plenty of time to work on it and I'm happy to do it, but I would really appreciate some guidance when I invariably get stuck a bit.  Again thanks for the help and direction.

**Comment by jadelord** [comment]
The CERN library has a Fortran77 implementations for all elliptic integrals. Among them there is a file on elliptic integral of the third kind, with a function named `RELI3` ([source code](http://cernlib.web.cern.ch/cernlib/download/2006_source/src/mathlib/gen/c/reli364.F)).

It is a translation of the ALGOL implementation found in: 

> R. BULIRSCH Numerical Calculation of Elliptic Integrals and  Elliptic Functions III.,  Numer. Math. 13 (1969) 305-315

Caution: Both the algorithms are quite complicated!

**TL;DR: ** According to [Lamarche and Leroy](http://dx.doi.org/10.1016/0010-4655(90)90184-3), it is possible to use existing `ellipkinc` and `ellipeinc` functions to evaluate **complete** elliptic integral of the third kind.

**Comment by pv** [comment]
cernlib license is [GPL](https://cernlib.web.cern.ch/cernlib/conditions.html), so the source code is not directly useful for us. The original paper however probably is...

**Comment by gmcastil** [comment]
This is still something on my radar screen, I'm just been really busy and I was going to wait until I have had a chance to learn C before working on this to address the comments made by @ewmoore.  I think it would be really cool to implement the Legendre forms in terms of the Carlson forms, and then support both.  Does anyone have a suggestion on where I might look to get an idea for how I might implement these as ufuncs as suggested?

**Comment by pv** [comment]
@gmcastil: it's fairly simple:
1) write the computational routine in C, C++, Fortran or Cython
2) add it in `generate_ufuncs.py` and `add_newdocs.py` and run `generate_ufuncs.py`
See eg https://github.com/scipy/scipy/pull/5931 or https://github.com/scipy/scipy/pull/5869 or other PRs doing this for examples

**Comment by gmcastil** [comment]
Wow, yeah that is a lot simpler than I was anticipating.  Thanks for some unambiguous direction.

**Comment by person142** [comment]
@gmcastil: since this is going to involve complex arithmetic, there are a couple of caveats which might save you some headaches later. Currently Scipy doesn't use C99 complex numbers, so if you want to code in C you're going to need to use numpy's ``cdouble``. (As of now) numpy doesn't expose complex arithmetic for ``cdouble``, so you'd have to roll your own arithmetic. It's easy to screw this up if you aren't careful, and you can forget about nice infix notation.

An alternative is to use Cython, though it's worth noting that (as of now) Cython's complex division is implemented naively and will overflow before it should. There's a function ``zdiv`` in ``_complexstuff.pxd`` to work around this. (There are plenty of situations, however, where this won't be a problem.)

If you code in Fortran (maybe @pv will want to correct me here) I believe you're restricted to using Fortran 77, which is not most people's idea of a good time.

I don't know anything about using C++.

**Comment by pv** [comment]
C++ can use std::complex; the code looks fairly much like one with C99 
double complex cf. the faddeeva implementation. IIRC, Fortran90 is also 
acceptable, mingw32 on Windows was the only blocker and we gave up using 
it for the last releases.

**Comment by arokem** [comment]
I see that this is still open, @rafaelnh implemented Carlson integrals in DIPY a while back: https://github.com/nipy/dipy/blob/master/dipy/reconst/dki.py#L49-L189, including tests that implement the equations from the Carlson paper that was mentioned earlier in this thread:  https://github.com/nipy/dipy/blob/master/dipy/reconst/tests/test_dki.py#L175-L243. Would it be useful to port this code over into scipy.special?

**Comment by sergiopasra** [comment]
Yes, please

**Comment by lucas-wilkins** [comment]
I was just trying to write something that needs elliptic functions of the third kind. I take it they're still not implemented in scipy?

**Comment by arokem** [comment]
I believe these were added in https://github.com/scipy/scipy/pull/15135 and now available here: https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.elliprj.html

**Comment by lucas-wilkins** [comment]
I saw that, I was a bit unsure as whether or not that particular function corresponds is a generalisation of what is often called the incomplete integral of the third kind (denoted Pi), which is a three, not four, parameter function. I can imagine it might be with the right substitutions, but its not obvious.

**Comment by bamford** [comment]
Looking at the [GSL code](https://git.savannah.gnu.org/cgit/gsl.git/tree/specfunc/ellint.c) and [`sympy` docs](https://docs.sympy.org/latest/modules/functions/special.html?highlight=elliptic#module-sympy.functions.special.elliptic_integrals), I've just implemented the elliptic integral of the third kind as follows:
```python
from scipy.special import elliprf, elliprj

def ellipp(n, m):
    if m >= 1:
        raise ValueError('m must be < 1')
    y = 1 - m
    rf = elliprf(0, y, 1)
    rj = elliprj(0, y, 1, 1 - n)
    return rf + rj * n / 3

def ellippinc(phi, n, m):
    nc = np.floor(phi / np.pi + 0.5)
    phi -= nc * np.pi
    sin_phi  = np.sin(phi)
    sin2_phi = sin_phi  * sin_phi;
    sin3_phi = sin2_phi * sin_phi;
    x = 1 - sin2_phi;
    y = 1 - m * sin2_phi;
    rf = elliprf(x, y, 1)
    rj = elliprj(x, y, 1, 1 - n * sin2_phi)
    val = sin_phi * rf + sin3_phi * rj * n / 3
    if nc != 0:
        rp = ellipp(n, m)
        val += 2 * nc * rp
    return val
```
The return values match those of `sympy.elliptic_pi` (but only the real part).
```python
ellipp(0.2, 0.7), sy.elliptic_pi(0.2, 0.7).evalf()
# (2.3590093147734037, 2.35900931477340)

ellippinc(0.3, 0.2, 0.7), sy.elliptic_pi(0.2, 0.3, 0.7).evalf()
# (0.30500346950008744, 0.305003469500087)

ellipp(2.2, 0.7), sy.elliptic_pi(2.2, 0.7).evalf()
# (-0.5104423797815891, -0.510442379781589 - 1.73658067934905*I)
```

Would it be worth me putting together a PR to include these in `scipy.special`?

Note that there are a couple of special cases alluded to (but not handled) in the GSL code.

**Comment by tupui** [comment]
Hi @bamford, I am sorry I had to close your PR. You linked GSL code which is GPL. This license is not compatible with SciPy and the only action of looking at the code is forbidden. See our documentation for more licensing information https://scipy.github.io/devdocs/dev/core-dev/index.html#licensing

**Comment by mdhaber** [comment]
Thanks @bamford. I thought it should also be mentioned that features of Boost are now relatively easy to add to SciPy, and its implementations were discussed above.

**Comment by duetosymmetry** [comment]
@bamford @tupui I agree that we should add documentation for how Carlson's `elliprj` can be used to compute Legendre's `ellippi` and `ellippiinc` (which I assume are the correct names to use, to match the conventions for the rest of the Legendre elliptic integrals). Including the info from some subset of https://dlmf.nist.gov/19.25.E2 , https://dlmf.nist.gov/19.25.E4 , https://dlmf.nist.gov/19.25.E14 , and https://dlmf.nist.gov/19.25.E16 is a minimum.

On the topic of a PR: @congma went very to try to ensure that his `ellipr{c,d,f,g,j}` functions would be valid in (much of) the complex plane. Since the same is possible for `ellippi` and `ellippiinc`, we should do that, too. Also, somebody told me that we should prefer to implement in C(++) and make a ufunc, rather than making a python wrapper that dispatches to `elliprj` and friends. But I'd be happy to hear otherwise.

**Comment by tupui** [comment]
> Also, somebody told me that we should prefer to implement in C(++) and make a ufunc, rather than making a python wrapper that dispatches to `elliprj` and friends. But I'd be happy to hear otherwise.

We don't have any hard requirement on using C(++)/Cython/Pythran or Python. Most of the time, it's actually better to start things with a Python version, and once it's polished and there is a **real** (insist on this) need for speed, then we can consider doing something else. We have more maintainers available for Python only things which is something to consider.

**Comment by bamford** [comment]
Just for a bit of context, I encountered the need for this function when providing feedback on a student's code, in which they were calculating the 3D magnetic field for a configuration of [solenoids](https://en.wikipedia.org/wiki/Solenoid#Finite_continuous_solenoid). As `ellippi` is not available in scipy they were using `sympy.elliptic_pi`. That was forcing them to loop over coordinates, which I assumed would be rather slow. I tried a few approaches to speeding it up, but kept coming back to wishing there was a numpy-compatible version of `ellippi`, so I looked into it, ended up with the code above (which was ~100 times faster than using sympy), and thought it might be a useful starting point for others.

It's been a very long time since I've done any complex analysis, so I don't think I'm going to be able to help with the finer points. However, it does seem like `ellippi` would be a useful addition to scipy.

**Comment by mdhaber** [comment]
I investigated the definition of the incomplete version of this function (presumably to be named `ellippiinc`) in terms of `elliprf` and `elliprj` as [Boost does](https://live.boost.org/doc/libs/1_38_0/libs/math/doc/sf_and_dist/html/math_toolkit/special/ellint/ellint_3.html)[^1]:

![image](https://github.com/user-attachments/assets/77c1081f-81e9-4aa4-936f-143a7ecbc261)

```python3
def pi(n, phi, m):
    return (sin(phi) * Rf(cos(phi)**2, 1-m*sin(phi)**2, 1)
            + n/3*sin(phi)**3 * Rj(cos(phi)**2, 1-m*sin(phi)**2, 1, 1-n*sin(phi)**2))
```

[^1]: Why roll our own if Boost has it? Boost's parameterization accepts a parameter $k$ where our $m = k^2$. So for Boost, the requirement is $-1 < k < 1$, but this would restrict our domain unnecessarily. If Boost were to provide better accuracy within its domain, we could use it where applicable, but we at least need to explore our own implementation for $m < -1$.

The domain of interest is:
- Real parameters only, in line with the other Legendre elliptic integral functions.
- $-\frac{\pi}{2} < \phi < \frac{\pi}{2}$ because $\phi$ only appears with the $\sin$ or $\cos$ function. Anything that goes wrong past these bounds is essentially a problem with `sin` or `cos`.
- $n < 1$, at least initially. Boost says the domain is restricted to "n < 1, otherwise returns the result of domain_error (outside this range the result would be complex)." I don't really see why that would happen. The real reason I'm restricting the attention there is that `mpmath`'s implementation starts returning complex numbers when $n > \frac{1}{\sin^2\phi}$, although I'll discuss this a bit more later.
- $m < \frac{1}{\sin^2\phi}$ because the result would indeed be complex, and in these cases, the other Legendre elliptic integral functions return NaN.

Starting with a moderate range of the parameter $n$ and $m = 0.5$, plot the $log_{10}$ of the relative error between this implementation (which I'll just call `pi`) and that of `mpmath` (which I'll call `pi_ref`).

<details>

```python3
import numpy as np
import matplotlib.pyplot as plt
from mpmath import mp
mp.dps = 50
rng = np.random.default_rng(1638083107694713882823079058616272161)
from scipy.special import elliprf as Rf, elliprj as Rj
from numpy import sin, cos

@np.vectorize
def pi_ref(n, phi, m):
    return complex(mp.ellippi(n, phi, m))

def pi(n, phi, m):
    return (sin(phi) * Rf(cos(phi)**2, 1-m*sin(phi)**2, 1)
            + n/3*sin(phi)**3 * Rj(cos(phi)**2, 1-m*sin(phi)**2, 1, 1-n*sin(phi)**2))

def log10_relative_error(res, ref):
    with np.errstate(all='ignore'):
        out = np.where(res != ref, abs(res - ref) / abs(ref), 0)
        out = np.log10(out)
    out[out < -16] = -16
    return out

def plot(n, phi, res, ref):
    logerr = log10_relative_error(res, ref)
    n, phi = n.ravel(), phi.ravel()
    plt.contourf(n, phi, logerr.T, cmap='viridis')
    plt.xlabel("$n$")
    plt.ylabel(r"$\phi$")
    plt.colorbar()
    plt.show()

phi = np.linspace(-np.pi/2, np.pi/2, 25)
n = np.linspace(-100, 1, 25)[:, np.newaxis]
m = 0.5

res = pi(n, phi, m)
ref = pi_ref(n, phi, m)
plot(n, phi, res, ref)
```
</details>

![image](https://github.com/user-attachments/assets/23814fdd-3f1f-4227-9447-e9c01ba6dde0)

I've produced many plots like this, manually exploring the space of $m$ and $n$ to some extent, and this sort of accuracy is typical.

We can also sample at random from the floating point numbers across the domain.

<details>

```python3
rng = np.random.default_rng(2184358452652)

types = {np.float16: np.int16, np.float32: np.int32, np.float64: np.int64,
         np.int16: np.float16, np.int32: np.float32, np.int64: np.float64}

def float2int(xf):
    """Convert the ith floating point number from zero to integer i"""
    # Adapted from numpy.testing._private.utils.integer_repr
    xf = np.asarray(xf)
    itype = types[xf.dtype.type]
    xi = xf.view(itype).copy()
    comp = np.asarray(np.iinfo(itype).min, dtype=itype)[()]
    xi[xi < 0] = comp - xi[xi < 0]
    return xi

def int2float(xi):
    """Convert integer i to the ith floating point number from zero"""
    # Adapted from numpy.testing._private.utils.integer_repr
    xi = np.asarray(xi).copy()
    comp = np.asarray(np.iinfo(xi.dtype).min, dtype=xi.dtype)[()]
    xi[xi < 0] = comp - xi[xi < 0]
    ftype = types[xi.dtype.type]
    xf = xi.view(ftype).copy()
    return xf

def sample(a, b, size, rng):
    """Sample uniformly from the set of floating point numbers between a and b"""
    int_range = float2int(np.asarray([a, b]))
    ints = rng.integers(*int_range, size=size)
    return int2float(ints)

rng = np.random.default_rng(6541968165162518)

n_points = 10000
phi = sample(-np.pi/2, np.pi/2, n_points, rng)
n = sample(-1e30, 1, n_points, rng)
m = sample(-1e30, 1e30, n_points, rng)

# filter out points where result would be complex
i_bad = m > 1/np.sin(phi)**2
phi = phi[~i_bad]
n = n[~i_bad]
m = m[~i_bad]

res = pi(n, phi, m)
ref = pi_ref(n, phi, m)
err = log10_relative_error(res, ref)
plt.hist(err, bins=64)
plt.xlabel('$\log_{10}$ relative error')
plt.ylabel('frequency')
plt.show()

i = err > -8
print(m[i], n[i], phi[i], ref[i])
# [ 4.34569731e-247 -3.97567079e-227] [-1.15440778e+26 -6.29851558e+27] [ 2.01024432e-05 -8.96069685e-05] [ 1.46197604e-13+0.j -1.97925054e-14+0.j]
```

</details>

![image](https://github.com/user-attachments/assets/8cae5491-15f3-497a-9164-89d02fc398e5)

With a more moderate domain, *linearly* sampling from $-100 < m < 100$ and $-100 < n < 100$ (avoiding either exceeding $\frac{1}{\sin^2 \phi}$ this looks like:

![image](https://github.com/user-attachments/assets/ed2ca10e-1ef4-4778-80d2-18c56ba2c26a)

With log-uniform values of `m` and `n` between `-1e30` and `-1`:

![image](https://github.com/user-attachments/assets/e3032cde-7c21-4119-8253-f3a45d18b5cd)

But here, all the reference values are less than `1e-8`, so the absolute error is never poor. 

So as far as I've seen, the relative error is always very good, and when it is not, the absolute error is OK. Perhaps there are other algorithms that could be used if the distinction between `3.47552775e-08` and `3.47552826e-08` is critical, but I think the way to start is to just document that the relative error of small absolute outputs may not be good.

**Edge case**
For $m > \frac{1}{\sin^2 \phi}$, the result would be complex, and `pi` naturally produces NaNs. (In `elliprj`, "Notably, x, y, and z must have non-negative real parts...") I think this is the desired behavior for now because that is what `ellipeinc` and `ellipkinc` do.

As far as I can tell, the behavior for $n >= \frac{1}{\sin^2 \phi}$ is OK in that the results produced by `pi` seem to agree with the value produced by Wolfram Alpha and _real_ part of the value produced by `pi_ref`:

```
pi(1, np.pi/2, 0.5)  # np.float64(nan)
pi_ref(1, np.pi/2, 0.5)[()]  # np.complex128(2.3095860183650104e+16+0j)
# EllipticPi[1, pi/2, 0.5]  # (complex infinity)

pi(1.001, np.pi/2, 0.5) # np.float64(-0.845750433748182)
pi_ref(1.001, np.pi/2, 0.5) # np.complex128(-0.8457504337481825-70.21308459432095j)
# EllipticPi[1.001, pi/2, 0.5]  # -0.8457504337482

pi(1000, 0.5, 0.5) # np.float64(0.001705833628135478)
pi_ref(1000, 0.5, 0.5) # np.complex128(0.0017058336281353885-0.0497102255527556j)
# EllipticPi[1000, pi/2, 0.5]  # 0.001705833628
```

It's just a little harder to exhaustively explore this region because `mpmath` slows down considerably, and I can't easily evaluate the integral numerically in this range, either, since the integrand goes through a singularity. It would probably not be wrong to just return `np.nan` here, though. Perhaps I'll dig a little deeper.
